cmd/argo/commands/copy_artifacts.go

```diff
@@ -0,0 +1,184 @@
+	"io"
+	"log"
+	"net/http"
+	"os"
+	"path/filepath"
+	"regexp"
+)
+
+type copyArtifactOps struct {
+	namespace    string // --namespace
+	workflowName string // --workflow-name
+	nodeId       string // --node-id
+	templateName string // --template-name
+	artifactName string // --artifact-name
+	artifactType string // --artifact-type
+}
+
+func NewCopyArtifactCommand() *cobra.Command {
+	var copyArtifactArgs copyArtifactOps
+	var argoServerArgs apiclient.ArgoServerOpts
+
+	command := &cobra.Command{
+		Use:   "cp outputDir ...",
+		Short: "copy artifacts from workflow",
+		Example: `# Copy a workflow's artifacts to a local output directory:
+
+  argo cp output-directory --workflow-name=my-wf
+
+# Copy artifacts from a specific node in a workflow to a local output directory:
+
+  argo cp output-directory --workflow-name=my-wf --node-id=my-wf-node-id-123
+`,
+		Run: func(cmd *cobra.Command, args []string) {
+			ctx, apiClient := client.NewAPIClient(cmd.Context())
+			serviceClient := apiClient.NewWorkflowServiceClient()
+			namespace := client.Namespace()
+			if len(copyArtifactArgs.namespace) > 0 {
+				namespace = copyArtifactArgs.namespace
+			}
+			workflow, err := serviceClient.GetWorkflow(ctx, &workflowpkg.WorkflowGetRequest{
+				Name:      copyArtifactArgs.workflowName,
+				Namespace: namespace,
+			})
+			if err != nil {
+				log.Fatal(err)
+			}
+
+			copyArtifactArgs.workflowName = workflow.Name
+
+			//{<node name1>: [artifact list], <node name2>: [artifact list],...}
+			artifactOutput := make(map[string]v1alpha1.Artifacts)
+			//var artifactOutput v1alpha1.Artifacts
+			//Getting a list of Artifacts based on user input
+			if len(copyArtifactArgs.nodeId) > 0 && len(copyArtifactArgs.templateName) > 0 {
+				log.Fatal("Only one of --template-name and --node-id can be filled out")
+			}
+			var nodes v1alpha1.Nodes
+			if len(copyArtifactArgs.nodeId) > 0 {
+				nodes = workflow.Status.Nodes.Filter(func(n v1alpha1.NodeStatus) bool { return n.ID == copyArtifactArgs.nodeId })
+			} else if len(copyArtifactArgs.templateName) > 0 {
+				nodes = workflow.Status.Nodes.Filter(func(n v1alpha1.NodeStatus) bool { return n.TemplateName == copyArtifactArgs.templateName })
+			} else {
+				nodes = workflow.Status.Nodes
+			}
+			for _, node := range nodes {
+				if node.Outputs != nil {
+					artifactOutput[node.ID] = node.Outputs.Artifacts
+				}
+			}
+
+			//Filtering in case of specifying artifact name
+			keysToDelete := make([]string, 0)
+			if len(copyArtifactArgs.artifactName) > 0 {
+				for key, artifactList := range artifactOutput {
+					artifactPtr := artifactList.GetArtifactByName(copyArtifactArgs.artifactName)
+					if artifactPtr != nil {
+						artifactOutput[key] = v1alpha1.Artifacts{*artifactPtr}
+					} else {
+						keysToDelete = append(keysToDelete, key)
+					}
+				}
+				for _, key := range keysToDelete {
+					delete(artifactOutput, key)
+				}
+			}
+
+			outputDir := args[0]
+			if len(outputDir) == 0 {
```

Comment:
meh - I would not check this - you have to write another test

Reply:
I removed the check, what should I verify for the test?